packages/snaps-utils/src/url.ts

```diff
@@ -0,0 +1,102 @@
+import { type SnapId } from '@metamask/snaps-sdk';
+import { assert } from '@metamask/utils';
+
+import { assertIsValidSnapId, stripSnapPrefix } from './snaps';
+
+export type Authority = 'client' | 'snap';
+
+export const CLIENT_PATHS = ['/'];
+
+export const SNAP_PATHS = ['/home'];
+
+/**
+ * Parse a url string to an object containing the authority, path and Snap id (if snap authority).
+ * This also validates the url before parsing it.
+ *
+ * Note: The Snap id returned from this function should always be validated to be an installed snap.
+ *
+ * @param str - The url string to validate and parse.
+ * @returns A parsed url object.
+ * @throws If the authority or path is invalid.
+ */
+export function parseMetaMaskUrl(str: string): {
+  authority: Authority;
+  snapId?: SnapId;
+  path: string;
+} {
+  const baseErrorMessage = 'Invalid MetaMask url:';
+  const url = new URL(str);
+  const { hostname: authority, pathname: path, protocol } = url;
+  if (protocol !== 'metamask:') {
+    throw new Error(
+      `Unable to parse URL. Expected the protocol to be "metamask:", but received "${protocol}".`,
+    );
+  }
+  switch (authority) {
+    case 'client':
+      assert(
+        CLIENT_PATHS.includes(path),
+        `Unable to navigate to "${path}". The provided path is not allowed.`,
+      );
+      return {
+        authority,
+        path,
+      };
+    case 'snap':
+      return parseSnapPath(path, baseErrorMessage);
+    default:
+      throw new Error(`${baseErrorMessage} invalid authority.`);
+  }
+}
+
+/**
+ * Parses a snap path and throws if it is invalid, returns an object with link data otherwise.
+ *
+ * @param path - The snap path to be parsed.
+ * @param baseErrorMessage - The base error message to throw.
+ * @returns A parsed url object.
+ * @throws If the path or Snap id is invalid.
+ */
+function parseSnapPath(
+  path: string,
+  baseErrorMessage: string,
+): {
+  authority: Authority;
+  snapId: SnapId;
+  path: string;
+} {
+  const strippedPath = stripSnapPrefix(path.slice(1));
+  const location = path.slice(1).startsWith('npm:') ? 'npm:' : 'local:';
+  const isNameSpaced = strippedPath.startsWith('@');
+  const pathTokens = strippedPath.split('/');
+  const lastPathToken = `/${pathTokens[pathTokens.length - 1]}`;
+  let partialSnapId;
+  if (location === 'local:') {
+    const [localProtocol, , ...rest] = pathTokens.slice(0, -1);
+    partialSnapId = `${localProtocol}//${rest.join('/')}`;
+    // we can't make assumptions of the structure of the local snap url since it can have a nested path
+    // so we only check that the last path token is one of the allowed paths
```

Comment:
What if the local snap ID is `local:http://localhost/home`? We didn't specify a path but are still being navigated to the home page 😅 I know this is an edge case but we should think about it at least.

Reply:
Yeah, I agree we should address edge cases for sure. In this case in particular, the navigation wouldn't happen because the reconstructed snap id would be `local:http://localhost`, it would pass the `assertIsValidSnapId` check but fail in `validateLink` when it's checked if that reconstructed snap id is actually installed. It might not throw the most informative error to the edge case but we can make a recommendation in our documentation to not serve local snaps from directories named as one of the snap routes.